lib/meandro/util.ex

```diff
@@ -16,27 +16,47 @@ defmodule Meandro.Util do
   It can be in `:parallel` or `:sequential` depending its second argument.
   """
   @spec parse_files([Path.t()], parsing_style()) :: [
-          {Path.t(), Macro.t()}
+          {Path.t(), [{module(), Macro.t()}]}
         ]
   def parse_files(paths, :sequential) do
-    Enum.map(paths, fn p ->
-      f = File.open!(p)
-      c = IO.read(f, :all)
-      ast = Code.string_to_quoted!(c)
-      {p, ast}
-    end)
+    paths
+    |> Enum.map(&file_to_ast/1)
+    |> List.flatten()
   end
 
   def parse_files(paths, :parallel) do
-    fun = fn p ->
-      f = File.open!(p)
-      c = IO.read(f, :all)
-      ast = Code.string_to_quoted!(c)
-      {p, ast}
-    end
-
     paths
-    |> Enum.map(&Task.async(fn -> fun.(&1) end))
+    |> Enum.map(&Task.async(fn -> file_to_ast(&1) end))
     |> Enum.map(&Task.await/1)
+    |> List.flatten()
+  end
+
+  defp file_to_ast(file) do
+    file
+    |> File.open!()
+    |> IO.read(:all)
+    |> Code.string_to_quoted!()
+    |> maybe_split_by_module(file)
+  end
+
+  defp maybe_split_by_module(ast, file) do
+    {_, result} = Macro.prewalk(ast, [], &collect_modules/2)
+    {file, result}
+  end
+
+  defp collect_modules({:defmodule, _, _} = module_node, acc) do
+    {module_node, [{module_name(module_node), module_node} | acc]}
```

Comment:
I believe we should also pop out nested module from the ast in order to not analyze them twice

Reply:
agreed